core/pipeline/pipeline.go

```diff
@@ -172,15 +206,19 @@ func (pipe *Pipeline) Run() *Context {
 			log.Debugf("%s-%s: %v", pipe.name, v.Name(), err)
 			pipe.context.Payload = err.Error()
 			pipe.handlePipelineError()
+			apm.CaptureError(apmSpanContext, err).Send()
+			apmSpan.End()
 			return pipe.context
 		}
 		log.Trace(pipe.name, ", end joint,", v.Name())
+		apmSpan.End()
 	}
 
 	return pipe.context
 }
 
 func (pipe *Pipeline) startPipeline() {
+	log := pipe.context.apmLogger
 
 	log.Trace("start pipeline: ", pipe.name)
```

Comment:
If I am following the changes correctly, it seems like the apmLogger can be nil. Because of this, adding nil checks or a method that can return a blank logger implementation in the event the apmLogger is not set,  would be needed for anyone using this module without an apmLogger set on the pipeline's context. This would also have the benefit of making these changes non breaking, ie the apmLogger is optional.

Reply:
Shoot, forgot to use the method version there. I added a function that grabs the logger and defaults to the current seelog logger if apm is null.